### PR TITLE
Create two bootloader programmers for gemmaBoot boards

### PR DESCRIPTION
**Pull request by stefanrueger**
Supposed to supersede PR #983

### CODE REVIEW

File: src/avrdude.conf.in [hunk 1]
```diff
@@ -1458,6 +1458,23 @@ programmer
...
 
+#------------------------------------------------------------
+# gemmaBoot
+#------------------------------------------------------------
+
+# Serving the Gemma bootloader
+# https://github.com/adafruit/Adafruit-Trinket-Gemma-Bootloader
+
+programmer
+    id                     = "gemmaBoot";
```
Comment (MCUdude) [inline src/avrdude.conf.in:1469]: Thanks for the PR! perhaps the id should be `"gemmaboot"` with a lowercase B instead? It looks like most programmers in avrdude.conf have id's in lowercase letters

**Comment by MCUdude** [comment]
According to the [Adafruit Gemma setup guide](https://learn.adafruit.com/introducing-gemma/setting-up-with-arduino-ide) The Gemma bootloader uses the same USB VID/PID as the USBtiny programmer. This means that the board has previously only been using the `-cusbtiny` option. Since `gemmaboot` will be a new USBtiny derived programmer, I think it would be best to be consistent with the names and use all lower case letters. Just my two cents. Waiting for @umbynos' response

File: src/avrdude.conf.in [hunk 2]
```diff
@@ -1458,6 +1458,23 @@ programmer
...
+    desc                   = "Gemma bootloader disguised as USBtiny";
+    type                   = "usbtiny";
+    prog_modes             = PM_SPM;
+    connection_type        = usb;
+    usbvid                 = 0x2341;
+    usbpid                 = 0x0105;
```
Comment (mcuee) [inline src/avrdude.conf.in:1475]: For Arduino Gemma Bootloader
usbpid = 0x2341
usbpid = 0x0c9f

For Adafruit Trinket Gemma Bootloader
usbpid = 0x1781
usbpid = 0x0c9f

File: src/avrdude.conf.in (hunk 1 above)
Comment (mcuee) [reply src/avrdude.conf.in:1469]: It is better to be consistent and use `gemmaboot`.

**Comment by stefanrueger** [comment]
Yes, it's like founding a rock band. 90% of the effort goes into choosing the name ;) But that's because this *is* important, of course.

OK, I see: we need *two* programmers for different boards? One with the name arduinogemma and one with the name adafruitgemma or some such? 

Best if you @MCUdude, @mcuee and @umbynos agree on the naming. Please push changes to avrdude.conf.in onto this branch. Neat if each programmer entry has an URL. Important that they have the `prog_modes = PM_SPM;` line in there (to indicate the programmer serves a bootloader). Super neat if you used the formatted output of `avrdude -c*gemma/s` to populate the PR. Once confirmed working by @umbynos someone can merge with the next batch of PRs (I'll be busy for most of Sept).

**Comment by mcuee** [comment]
> OK, I see: we need two programmers for different boards? One with the name arduinogemma and one with the name adafruitgemma or some such?

@stefanrueger 
Yes that is correct. I will vote for the name `arduino_gemma` and `adafruit_gemma`. But I am okay with other names that @MCUdude and @umbynos like.

```
#------------------------------------------------------------
# arduino_gemma
#------------------------------------------------------------

# https://github.com/arduino/ArduinoCore-avr/tree/master/bootloaders/gemma

programmer
    id                     = "arduino_gemma";
    desc                   = "Arduino Gemma bootloader disguised as USBtiny";
    type                   = "usbtiny";
    prog_modes             = PM_SPM;
    connection_type        = usb;
    usbvid                 = 0x2341;
    usbpid                 = 0x0c9f;
;

#------------------------------------------------------------
# adafruit_gemma
#------------------------------------------------------------

# Serving the Gemma bootloader
# https://github.com/adafruit/Adafruit-Trinket-Gemma-Bootloader

programmer
    id                     = "adafruit_gemma";
    desc                   = "Adafruit Trinket Gemma bootloader disguised as USBtiny";
    type                   = "usbtiny";
    prog_modes             = PM_SPM;
    connection_type        = usb;
    usbvid                 = 0x1781;
    usbpid                 = 0x0c9f;
;
```

**Comment by umbynos** [comment]
>Yes that is correct. I will vote for the name arduino_gemma and adafruit_gemma

Ok for me

>Once confirmed working by @umbynos someone can merge with the next batch of PRs

I'll try to find some time to test this out in the next couple of days. In the meantime, thanks a lot for your work

**Comment by mcuee** [comment]
Somehow it does not work for my ATtiny85 breakout board. It seems something is wrong with usbtiny support.

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1093 -C .\avrdude_pr1093.conf -c arduino_gemma -p t85 -U .\Blink.ino.gemma.hex


avrdude_pr1093.exe: error: usbtiny_transmit: Input/output error
avrdude_pr1093.exe: initialization failed, rc=-1
                    Double check connections and try again, or use -F to override
                    this check.


avrdude_pr1093.exe: error: usbtiny_transmit: Input/output error

avrdude_pr1093.exe done.  Thank you.

```

Arduino is okay and it uses a gemma specific avrdude.conf.
```
C:\Users\xiaof\AppData\Local\Arduino15\packages\arduino\tools\avrdude\6.3.0-arduino17/bin/avrdude
 -CC:\Users\xiaof\AppData\Local\Arduino15\packages\arduino\hardware\avr\1.8.5/bootloaders/gemma/avrdude.conf
 -v -pattiny85 -carduinogemma -Uflash:w:C:\Users\xiaof\AppData\Local\Temp\arduino_build_742517/Blink.ino.hex:i 

avrdude: Version 6.3-20190619
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "C:\Users\xiaof\AppData\Local\Arduino15\packages\arduino\hardware\avr\1.8.5/bootloaders/gemma/avrdude.conf"

         Using Port                    : usb
         Using Programmer              : arduinogemma
avrdude: usbdev_open(): Found USBtinyISP, bus:device: bus-0:\\.\libusb0-0001--0x2341-0x0c9f
         AVR Part                      : ATtiny85
         Chip Erase delay              : 400000 us
         PAGEL                         : P00
         BS2                           : P00
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    12     4    0 no        512    4      0  4000  4500 0xff 0xff
           flash         65     6    32    0 yes      8192   64    128 30000 30000 0xff 0xff
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00
           lock           0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           lfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           calibration    0     0     0    0 no          2    0      0     0     0 0x00 0x00

         Programmer Type : USBtiny
         Description     : Arduino Gemma. Based on USBtiny
avrdude: programmer operation not supported

avrdude: Using SCK period of 10 usec
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e930b (probably t85)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: Using SCK period of 10 usec
avrdude: reading input file "C:\Users\xiaof\AppData\Local\Temp\arduino_build_742517/Blink.ino.hex"
avrdude: writing flash (682 bytes):

Writing | ####avrdude: 3 retries during SPI command
#####avrdude: 4 retries during SPI command
####avrdude: 5 retries during SPI command
#####avrdude: 5 retries during SPI command
####avrdude: 5 retries during SPI command
#####avrdude: 4 retries during SPI command
####avrdude: 5 retries during SPI command
#####avrdude: 5 retries during SPI command
####avrdude: 4 retries during SPI command
#####avrdude: 5 retries during SPI command
##### | 100% 0.49s

avrdude: 682 bytes of flash written
avrdude: verifying flash memory against C:\Users\xiaof\AppData\Local\Temp\arduino_build_742517/Blink.ino.hex:
avrdude: load data flash data from input file C:\Users\xiaof\AppData\Local\Temp\arduino_build_742517/Blink.ino.hex:
avrdude: input file C:\Users\xiaof\AppData\Local\Temp\arduino_build_742517/Blink.ino.hex contains 682 bytes
avrdude: reading on-chip flash data:

Reading | ################################################## | 100% 0.07s

avrdude: verifying ...
avrdude: 682 bytes of flash verified

avrdude done.  Thank you.

```

Arduino 7.0 from Arduino is also okay with the gemma specific avrdude.conf.
```
PS C:\work\avr\avrdude_test\avrdude_bin\bin_misc\avrdude_7.0-arduino.3_Windows_32bit\avrdude_Windows_32bit\bin> 
.\avrdude -C .\gemma\avrdude.conf -c arduinogemma -p t85 -U .\Blink.ino.gemma.hex

avrdude.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude.exe: Device signature = 0x1e930b (probably t85)
avrdude.exe: NOTE: "flash" memory has been specified, an erase cycle will be performed
             To disable this feature, specify the -D option.
avrdude.exe: erasing chip
avrdude.exe: reading input file ".\Blink.ino.gemma.hex"
avrdude.exe: input file .\Blink.ino.gemma.hex auto detected as Intel Hex
avrdude.exe: writing flash (682 bytes):

Writing | #####                                              | 9% 0.04savrdude.exe: 5 retries during SPI command
Writing | #########                                          | 18% 0.09savrdude.exe: 7 retries during SPI command
Writing | ##############                                     | 27% 0.14savrdude.exe: 7 retries during SPI command
Writing | ##################                                 | 36% 0.19savrdude.exe: 7 retries during SPI command
Writing | #######################                            | 45% 0.24savrdude.exe: 7 retries during SPI command
Writing | ###########################                        | 54% 0.29savrdude.exe: 7 retries during SPI command
Writing | ################################                   | 63% 0.34savrdude.exe: 7 retries during SPI command
Writing | ####################################               | 72% 0.39savrdude.exe: 6 retries during SPI command
Writing | #########################################          | 81% 0.44savrdude.exe: 7 retries during SPI command
Writing | #############################################      | 90% 0.49savrdude.exe: 7 retries during SPI command
Writing | ################################################## | 100% 0.54s

avrdude.exe: 682 bytes of flash written
avrdude.exe: verifying flash memory against .\Blink.ino.gemma.hex:
avrdude.exe: input file .\Blink.ino.gemma.hex auto detected as Intel Hex

Reading | ################################################## | 100% 0.05s

avrdude.exe: 682 bytes of flash verified

avrdude.exe done.  Thank you.

```

**Comment by mcuee** [comment]
It seems to me my breakout board is not so stable with the libusb-win32 driver, switching to libusbk driver seems to fix the issue.

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1093 -C .\avrdude_pr1093.conf -c arduino_gemma -p t85 -U .\Blink.ino.gemma.hex -v

avrdude_pr1093.exe: Version 7.0-20220831 (a536a01)
                    Copyright (c) Brian Dean, http://www.bdmicro.com/
                    Copyright (c) Joerg Wunsch

                    System wide configuration file is ".\avrdude_pr1093.conf"

avrdude_pr1093.exe: input file .\Blink.ino.gemma.hex auto detected as Intel Hex
                    Using Port                    : usb
                    Using Programmer              : arduino_gemma
avrdude_pr1093.exe: usbdev_open(): Found USBtinyISP, bus:device: 001:031
                    AVR Part                      : ATtiny85
                    Chip Erase delay              : 4500 us
                    RESET disposition             : possible i/o
                    RETRY pulse                   : SCK
                    Serial program mode           : yes
                    Parallel program mode         : yes
                    Timeout                       : 200
                    StabDelay                     : 100
                    CmdexeDelay                   : 25
                    SyncLoops                     : 32
                    PollIndex                     : 3
                    PollValue                     : 0x53
                    Memory Detail                 :

                                                      Block Poll               Page                       Polled
                      Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
                      ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
                      eeprom                 65     6     4    0 no        512    4      0  4000  4500 0xff 0xff
                      flash                  65     6    32    0 yes      8192   64    128  4500  4500 0xff 0xff
                      lfuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                      hfuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                      efuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                      lock                    0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                      signature               0     0     0    0 no          3    1      0     0     0 0x00 0x00
                      calibration             0     0     0    0 no          1    1      0     0     0 0x00 0x00

                    Programmer Type : USBtiny
                    Description     : Arduino Gemma bootloader disguised as USBtiny
avrdude_pr1093.exe: programmer operation not supported

avrdude_pr1093.exe: Using SCK period of 10 usec
avrdude_pr1093.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude_pr1093.exe: Device signature = 0x1e930b (probably t85)
avrdude_pr1093.exe: NOTE: "flash" memory has been specified, an erase cycle will be performed
                    To disable this feature, specify the -D option.
avrdude_pr1093.exe: erasing chip
avrdude_pr1093.exe: Using SCK period of 10 usec
avrdude_pr1093.exe: reading input file .\Blink.ino.gemma.hex for flash
                    with 682 bytes in 1 section within [0, 0x2a9]
                    using 11 pages and 22 pad bytes
avrdude_pr1093.exe: writing 682 bytes flash ...

Writing | #####                                              | 9% 0.01savrdude_pr1093.exe: 5 retries during SPI command
Writing | #########                                          | 18% 0.03savrdude_pr1093.exe: 7 retries during SPI command
Writing | ##############                                     | 27% 0.05savrdude_pr1093.exe: 7 retries during SPI command
Writing | ##################                                 | 36% 0.07savrdude_pr1093.exe: 7 retries during SPI command
Writing | #######################                            | 45% 0.10savrdude_pr1093.exe: 7 retries during SPI command
Writing | ###########################                        | 54% 0.12savrdude_pr1093.exe: 7 retries during SPI command
Writing | ################################                   | 63% 0.14savrdude_pr1093.exe: 6 retries during SPI command
Writing | ####################################               | 72% 0.16savrdude_pr1093.exe: 7 retries during SPI command
Writing | #########################################          | 81% 0.18savrdude_pr1093.exe: 7 retries during SPI command
Writing | #############################################      | 90% 0.21savrdude_pr1093.exe: 7 retries during SPI command
Writing | ################################################## | 100% 0.23s

avrdude_pr1093.exe: 682 bytes of flash written
avrdude_pr1093.exe: verifying flash memory against .\Blink.ino.gemma.hex

Reading | ################################################## | 100% 0.06s

avrdude_pr1093.exe: 682 bytes of flash verified

avrdude_pr1093.exe done.  Thank you.


```

**Comment by stefanrueger** [comment]
@umbynos Thanks for checking and confirming that the change works with @mcuee's  programmer definitions. 

Can I suggest @mcuee pushes a commit where his programmer definitions replace the gemmaBoot one to the [gemmaBoot](https://github.com/stefanrueger/avrdude/tree/gemmaBoot) branch on [stefanrueger/avrdude](https://github.com/stefanrueger/avrdude). I am hoping the permissions allow this...

@MCUdude That's a good point. Could it be that other `usleep()` calls in `usbtiny.c` need increased sleep times for a bootloader to reflect the communication overhead? Might be a good to change as well if so. Don't have the devices (and am actually away from my usual setup) so best if @umbynos investigates?

Once we are all happy with this, I would be OK with merging this PR during the next merge fest in a few weeks' time

**Comment by mcuee** [comment]
> Can I suggest @mcuee pushes a commit where his programmer definitions replace the [gemmaBoot](https://github.com/stefanrueger/avrdude/tree/gemmaBoot) one to the gemmaBoot branch on [stefanrueger/avrdude](https://github.com/stefanrueger/avrdude). I am hoping the permissions allow this...

I believe I do not have the permissions to do that. 

@MCUdude Not so sure if you have the permissions or not. Please try that.

**Comment by mcuee** [comment]
> @MCUdude That's a good point. Could it be that other usleep() calls in usbtiny.c need increased sleep times for a bootloader to reflect the communication overhead? Might be a good to change as well if so. Don't have the devices (and am actually away from my usual setup) so best if @umbynos investigates?

Indeed. I was thinking my setup is not so stable (just a simple ATtiny85 breakout board with an LED) when I saw all the retries. But then @umbynos' results are also the same (using this pull request or using #983).

**Comment by stefanrueger** [comment]
> permissions

@mcuee I tried to solve this by inviting collaborators to my fork of AVRDUDE, but this is not the right tool, so I've cancelled the invitations. Apologies for the flurry of emails you (and @MCUdude and @dl8dtl) have received.

**Comment by stefanrueger** [comment]
Have now added the programmers by @mcuee to avrdude.conf.in myself. Sorry for my inability to figure out how to change permissions and apologies for the lack of attribution  to @mcuee for this commit (the main reason why I initially suggested @mcuee push the commit).

> retries during a write session

An interesting problem. I suspect this has to do with the fact that usbtiny has no handshake, so the usbtiny programmer might need to sleep longer in some places when talking to a bootloader?

**Comment by mcuee** [comment]
> > permissions
> 
> @mcuee I tried to solve this by inviting collaborators to my fork of AVRDUDE, but this is not the right tool, so I've cancelled the invitations. Apologies for the flurry of emails you (and @MCUdude and @dl8dtl) have received.

No problems.

**Comment by umbynos** [comment]
This is with [7.0-arduino.3](https://github.com/arduino/avrdude-packing/releases/tag/7.0-arduino.3) (so without this PR but with #983). I only had to add [this](https://github.com/arduino/ArduinoCore-avr/blob/master/bootloaders/gemma/avrdude.conf#L751-L758) to the `avrdude.config` and the arduino gemma:

```
"/home/umberto/.arduino15/packages/arduino/tools/avrdude/7.0-arduino.3/bin/avrdude" "-C/home/umberto/.arduino15/packages/arduino/tools/avrdude/7.0-arduino.3/etc/avrdude.conf" -v -V -pattiny85 -carduinogemma  "-Uflash:w:/tmp/arduino-sketch-DA9372EC17C7CC77521592697B1C7CD7/Blink.ino.hex:i"

avrdude: Version 7.0
         Copyright (c) Brian Dean, http://www.bdmicro.com/
         Copyright (c) Joerg Wunsch

         System wide configuration file is "/home/umberto/.arduino15/packages/arduino/tools/avrdude/7.0-arduino.3/etc/avrdude.conf"
         User configuration file is "/home/umberto/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : usb
         Using Programmer              : arduinogemma
avrdude: usbdev_open(): Found USBtinyISP, bus:device: 001:023
         AVR Part                      : ATtiny85
         Chip Erase delay              : 400000 us
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         Serial program mode           : yes
         Parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom                 65    12     4    0 no        512    4      0  4000  4500 0xff 0xff
           flash                  65     6    32    0 yes      8192   64    128 30000 30000 0xff 0xff
           signature               0     0     0    0 no          3    0      0     0     0 0x00 0x00
           lock                    0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           lfuse                   0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           hfuse                   0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           efuse                   0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           calibration             0     0     0    0 no          1    0      0     0     0 0x00 0x00

         Programmer Type : USBtiny
         Description     : Arduino Gemma. Based on USBtiny
avrdude: programmer operation not supported

avrdude: Using SCK period of 10 usec
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e930b (probably t85)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: Using SCK period of 10 usec
avrdude: reading input file "/tmp/arduino-sketch-DA9372EC17C7CC77521592697B1C7CD7/Blink.ino.hex"
avrdude: writing flash (682 bytes):

Writing | #####                                              | 9% 0.03savrdude: 9 retries during SPI command
Writing | #########                                          | 18% 0.07savrdude: 8 retries during SPI command
Writing | ################################                   | 63% 0.24savrdude: 9 retries during SPI command
Writing | ####################################               | 72% 0.28savrdude: 8 retries during SPI command
Writing | #########################################          | 81% 0.32savrdude: 8 retries during SPI command
Writing | #############################################      | 90% 0.36savrdude: 9 retries during SPI command
Writing | ################################################## | 100% 0.40s

avrdude: 682 bytes of flash written

avrdude done.  Thank you.
```

While this is with the old avrdude 6.3.0-arduino17, the version we are currently using in the latest version of the avr core.
```
"/home/umberto/.arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude" "-C/home/umberto/.arduino15/packages/arduino/hardware/avr/1.8.5/bootloaders/gemma/avrdude.conf" -v -V -pattiny85 -carduinogemma  "-Uflash:w:/tmp/arduino-sketch-DA9372EC17C7CC77521592697B1C7CD7/Blink.ino.hex:i"

avrdude: Version 6.3-20190619
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/home/umberto/.arduino15/packages/arduino/hardware/avr/1.8.5/bootloaders/gemma/avrdude.conf"
         User configuration file is "/home/umberto/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : usb
         Using Programmer              : arduinogemma
avrdude: usbdev_open(): Found USBtinyISP, bus:device: 001:024
         AVR Part                      : ATtiny85
         Chip Erase delay              : 400000 us
         PAGEL                         : P00
         BS2                           : P00
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    12     4    0 no        512    4      0  4000  4500 0xff 0xff
           flash         65     6    32    0 yes      8192   64    128 30000 30000 0xff 0xff
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00
           lock           0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           lfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  9000  9000 0x00 0x00
           calibration    0     0     0    0 no          2    0      0     0     0 0x00 0x00

         Programmer Type : USBtiny
         Description     : Arduino Gemma. Based on USBtiny
avrdude: programmer operation not supported

avrdude: Using SCK period of 10 usec
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e930b (probably t85)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: Using SCK period of 10 usec
avrdude: reading input file "/tmp/arduino-sketch-DA9372EC17C7CC77521592697B1C7CD7/Blink.ino.hex"
avrdude: writing flash (682 bytes):

Writing | #########                                          | 18% 0.07savrdude: 8 retries during SPI command
Writing | ##############                                     | 27% 0.10savrdude: 9 retries during SPI command
Writing | ##################                                 | 36% 0.14savrdude: 9 retries during SPI command
Writing | #######################                            | 45% 0.18savrdude: 9 retries during SPI command
Writing | ###########################                        | 54% 0.22savrdude: 8 retries during SPI command
Writing | ################################                   | 63% 0.26savrdude: 9 retries during SPI command
Writing | ####################################               | 72% 0.30savrdude: 9 retries during SPI command
Writing | #########################################          | 81% 0.34savrdude: 9 retries during SPI command
Writing | #############################################      | 90% 0.37savrdude: 9 retries during SPI command
Writing | ################################################## | 100% 0.41s

avrdude: 682 bytes of flash written

avrdude done.  Thank you.
```

> Is it normal to have this many retries during a write session?

I do see the retries also there... So I would say it's normal. But if we could solve them it would be super awesome!

>  need increased sleep times for a bootloader to reflect the communication overhead? Might be a good to change as well if so. Don't have the devices (and am actually away from my usual setup) so best if @umbynos investigates?

@stefanrueger where do you suggest starting the investigation? Unfortunately, I do not know the codebase at all

**Comment by mcuee** [comment]
> @stefanrueger where do you suggest starting the investigation? Unfortunately, I do not know the codebase at all.

I am guessing here. Since the SPI retry happens i the Flash Write stage, maybe we can zoom into the `usbtiny_paged_write()` function.
https://github.com/avrdudes/avrdude/blob/main/src/usbtiny.c#L707-L771

Especially the following.
https://github.com/avrdudes/avrdude/blob/main/src/usbtiny.c#L756-L758
```
		32 * PDATA(pgm)->sck_period + delay  // each byte gets turned into a
	                             // 4-byte SPI cmd usb_out() multiplies
	                             // this per byte. Then add the cmd-delay
```

Maybe increasing the cmd delay can help.

**Comment by mcuee** [comment]
The SPI delay can be further reduced.

```
xiaof@LAPTOP-913SIMUD MINGW64 /c/work/avr/avrdude_test/avrdude_sr
$ git diff
diff --git a/src/usbtiny.c b/src/usbtiny.c
index 7153ede..e822b7f 100644
--- a/src/usbtiny.c
+++ b/src/usbtiny.c
@@ -759,13 +759,13 @@ static int usbtiny_paged_write(const PROGRAMMER *pgm, const AVRPART *p, const AV
                addr,           // Address in memory
                m->buf + addr,  // Pointer to data
                chunk,          // Number of bytes to write
-               32 * PDATA(pgm)->sck_period + delay  // each byte gets turned into a
+               32 * PDATA(pgm)->sck_period + delay +5000 // each byte gets turned into a
                                     // 4-byte SPI cmd usb_out() multiplies
                                     // this per byte. Then add the cmd-delay
                ) < 0) {
       return -1;
     }
-
+       usleep(50000);
     next = addr + chunk;       // Calculate what address we're at now
     if (m->paged
        && ((next % page_size) == 0 || next == maxaddr) ) {

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1093_mod1 -C .\avrdude_pr1093.conf 
-c arduino_gemma -p t85 -U .\Blink.ino.gemma.hex

avrdude_pr1093_mod1.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude_pr1093_mod1.exe: Device signature = 0x1e930b (probably t85)
avrdude_pr1093_mod1.exe: NOTE: "flash" memory has been specified, an erase cycle will be performed
                         To disable this feature, specify the -D option.
avrdude_pr1093_mod1.exe: erasing chip
avrdude_pr1093_mod1.exe: reading input file .\Blink.ino.gemma.hex for flash
avrdude_pr1093_mod1.exe: writing 682 bytes flash ...

Writing | #####                                              | 9% 0.07savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | #########                                          | 18% 0.15savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | ##############                                     | 27% 0.24savrdude_pr1093_mod1.exe: 5 retries during SPI command
Writing | ##################                                 | 36% 0.32savrdude_pr1093_mod1.exe: 5 retries during SPI command
Writing | #######################                            | 45% 0.40savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | ###########################                        | 54% 0.48savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | ################################                   | 63% 0.56savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | ####################################               | 72% 0.64savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | #########################################          | 81% 0.72savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | #############################################      | 90% 0.80savrdude_pr1093_mod1.exe: 4 retries during SPI command
Writing | ################################################## | 100% 0.88s

avrdude_pr1093_mod1.exe: 682 bytes of flash written
avrdude_pr1093_mod1.exe: verifying flash memory against .\Blink.ino.gemma.hex

Reading | ################################################## | 100% 0.06s

avrdude_pr1093_mod1.exe: 682 bytes of flash verified

avrdude_pr1093_mod1.exe done.  Thank you.

```

**Comment by mcuee** [comment]
More proper patch.

```
$ git diff
diff --git a/src/usbtiny.c b/src/usbtiny.c
index 7153ede..8aab0b0 100644
--- a/src/usbtiny.c
+++ b/src/usbtiny.c
@@ -596,7 +596,8 @@ static int usbtiny_spi(const PROGRAMMER *pgm, const unsigned char *cmd, unsigned
   for (i = 0; i < count; i += 4) {
     if (usbtiny_cmd(pgm, cmd + i, res + i) < 0) {
       return -1;
-    }
+    if(pgm->prog_modes & PM_SPM) usleep(5000);
+       }
   }
   return 0;
 }
@@ -765,7 +766,7 @@ static int usbtiny_paged_write(const PROGRAMMER *pgm, const AVRPART *p, const AV
                ) < 0) {
       return -1;
     }
-
+       if(pgm->prog_modes & PM_SPM) usleep(50000);
     next = addr + chunk;       // Calculate what address we're at now
     if (m->paged
        && ((next % page_size) == 0 || next == maxaddr) ) {

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1093_mod3 -C .\avrdude_pr1093.conf 
-c arduino_gemma -p t85 -U .\Blink.ino.gemma.hex

avrdude_pr1093_mod3.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude_pr1093_mod3.exe: Device signature = 0x1e930b (probably t85)
avrdude_pr1093_mod3.exe: NOTE: "flash" memory has been specified, an erase cycle will be performed
                         To disable this feature, specify the -D option.
avrdude_pr1093_mod3.exe: erasing chip
avrdude_pr1093_mod3.exe: reading input file .\Blink.ino.gemma.hex for flash
avrdude_pr1093_mod3.exe: writing 682 bytes flash ...

Writing | #####                                              | 9% 0.07savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | #########                                          | 18% 0.15savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | ##############                                     | 27% 0.24savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | ##################                                 | 36% 0.32savrdude_pr1093_mod3.exe: 5 retries during SPI command
Writing | #######################                            | 45% 0.40savrdude_pr1093_mod3.exe: 5 retries during SPI command
Writing | ###########################                        | 54% 0.48savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | ################################                   | 63% 0.56savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | ####################################               | 72% 0.64savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | #########################################          | 81% 0.72savrdude_pr1093_mod3.exe: 5 retries during SPI command
Writing | #############################################      | 90% 0.80savrdude_pr1093_mod3.exe: 4 retries during SPI command
Writing | ################################################## | 100% 0.88s

avrdude_pr1093_mod3.exe: 682 bytes of flash written
avrdude_pr1093_mod3.exe: verifying flash memory against .\Blink.ino.gemma.hex

Reading | ################################################## | 100% 0.05s

avrdude_pr1093_mod3.exe: 682 bytes of flash verified

avrdude_pr1093_mod3.exe done.  Thank you.


```

**Comment by mcuee** [comment]
> More proper patch.
> 
> ```
> $ git diff
> diff --git a/src/usbtiny.c b/src/usbtiny.c
> index 7153ede..8aab0b0 100644
> --- a/src/usbtiny.c
> +++ b/src/usbtiny.c
> @@ -596,7 +596,8 @@ static int usbtiny_spi(const PROGRAMMER *pgm, const unsigned char *cmd, unsigned
>    for (i = 0; i < count; i += 4) {
>      if (usbtiny_cmd(pgm, cmd + i, res + i) < 0) {
>        return -1;
> -    }
> +    if(pgm->prog_modes & PM_SPM) usleep(5000);
> +       }
>    }
>    return 0;
>  }
```

Since reading is okay, maybe this is only needed for `usbtiny_paged_write()`.

**Comment by mcuee** [comment]
@stefanrueger 
I think you can merge this first and then we can further debug the SPI command retry count issue once you have some time later.

I will create another issue to capture the SPI retry count issue.

**Pull request merged**